packages/block-editor/src/components/list-view/private-list-view.js

```diff
@@ -0,0 +1,270 @@
+
+/**
+ * Internal dependencies
+ */
+import { store as blockEditorStore } from '../../store';
+
+import { BlockSettingsDropdown } from '../block-settings-menu/block-settings-dropdown';
+
+import ListViewBranch from './branch';
+import { ListViewContext } from './context';
+import ListViewDropIndicator from './drop-indicator';
+import useBlockSelection from './use-block-selection';
+import useListViewClientIds from './use-list-view-client-ids';
+import useListViewDropZone from './use-list-view-drop-zone';
+import useListViewExpandSelectedItem from './use-list-view-expand-selected-item';
+
+const expanded = ( state, action ) => {
+	if ( Array.isArray( action.clientIds ) ) {
+		return {
+			...state,
+			...action.clientIds.reduce(
+				( newState, id ) => ( {
+					...newState,
+					[ id ]: action.type === 'expand',
+				} ),
+				{}
+			),
+		};
+	}
+	return state;
+};
+
+export const BLOCK_LIST_ITEM_HEIGHT = 36;
+
+/** @typedef {import('react').ComponentType} ComponentType */
+/** @typedef {import('react').Ref<HTMLElement>} Ref */
+
+/**
+ * Show a hierarchical list of blocks.
+ *
+ * @param {Object}         props                         Components props.
+ * @param {string}         props.id                      An HTML element id for the root element of ListView.
+ * @param {Array}          props.blocks                  _deprecated_ Custom subset of block client IDs to be used instead of the default hierarchy.
+ * @param {?boolean}       props.showBlockMovers         Flag to enable block movers. Defaults to `false`.
+ * @param {?boolean}       props.isExpanded              Flag to determine whether nested levels are expanded by default. Defaults to `false`.
+ * @param {?boolean}       props.showAppender            Flag to show or hide the block appender. Defaults to `false`.
+ * @param {?ComponentType} props.blockSettingsMenu       Optional more menu substitution. Defaults to the standard `BlockSettingsDropdown` component.
+ * @param {string}         props.rootClientId            The client id of the root block from which we determine the blocks to show in the list.
+ * @param {string}         props.description             Optional accessible description for the tree grid component.
+ * @param {?Function}      props.onSelect                Optional callback to be invoked when a block is selected. Receives the block object that was selected.
+ * @param {Function}       props.renderAdditionalBlockUI Function that renders additional block content UI.
+ * @param {Ref}            ref                           Forwarded ref
+ */
+const PrivateListView = forwardRef( function ListViewComponent(
+	{
+		id,
+		blocks,
+		showBlockMovers = false,
+		isExpanded = false,
+		showAppender = false,
+		blockSettingsMenu: BlockSettingsMenu = BlockSettingsDropdown,
+		rootClientId,
+		description,
+		onSelect,
+		renderAdditionalBlockUI,
+	},
+	ref
+) {
+	// This can be removed once we no longer need to support the blocks prop.
+	if ( blocks ) {
```

Comment:
Are we sure we want to remove this? @talldan doesn't this remain a viable alternative, giving ListView a list of blocks? I think we're corneting ListView intro a place where it can only work along with an instanced block editor tree. I am suspicious if this actually improves ListView or takes away.

I'd be happy in the future to have some need to pass a tree of blocks and for it to work.

Reply:
I'm not sure when `blocks` was introduced, but it is a little open to issues:
- Some selectors that return block trees only contain uncontrolled blocks
- ListView is optimized to work with a `clientIdsTree` rather than a full block tree, passing in a full block tree is overly expensive.

> I think we're corneting ListView intro a place where it can only work along with an instanced block editor tree.

It's a bit of an existential issue. I'm not sure what use cases there would be for it work with a non-instanced block editor.